### PR TITLE
Show coloured tree-diff output in ImpTests

### LINKED ISSUE #4471: Show coloured tree-diff output in ImpTests
- [x] Change `logEntry` to take `Doc AnsiStyle` instead of `String`
- [x] Find a way to pass coloured text in an `HUnitFailure` (which is what gets thrown by `hpsec`)
- [x] Show complete stack trace in `logWithCallStack`
- [x] Improve the pretty-printing of cost models

### PR DESCRIPTION
**Pull request by neilmayhew**
# Description

We use `tree-diff` to show the difference between actual and expected output in tests, and although `tree-diff` very helpfully colours the insertions and deletions the output is currently entirely red.

There are two main reasons for this:

1. `hspec` forces a red colour on every line when it outputs the text of a failure message.

    This can be overcome by adding `--no-color` to the command-line test options. Unfortunately, this also drops using colour in the progress messages, but in most circumstances that's an acceptable tradeoff. When it's fairly certain that all tests will pass (eg when doing a final test run before pushing) omitting `--no-color` can be helpful.

    More investigation is needed to determine whether there's a way to defeat `hspec`'s aggressive colourization of failure messages.

1. The messages are rendered to `String` in order to become expectation failures, and this throws away all of the text attributes such as colour.

    This can be overcome by changing the method of rendering the `Doc AnsiStyle` produced by `tree-diff` so that the escape codes are embedded in the failure message.

    At the same time, it's also helpful to allow our own `ImpTest` logging functions to accept `Doc AnsiStyle` instead of `String` so that coloured text can be output to the log.

This PR makes the changes necessary for (2), and few other quality-of-life improvements related to the use of colour.

Closes #4471

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

**Comment by neilmayhew** [comment]
This is based on the branch from #4440 because otherwise they would conflict with each other. I'm assuming that #4440 will be merged first.

**Comment by neilmayhew** [comment]
We may want to include changing the CI workflow to use `--no-color`

**Review by Soupstraw — APPROVED** [review]
Beautiful

File: .github/workflows/haskell.yml [hunk 1]
```diff
@@ -326,11 +326,21 @@ jobs:
     - name: Run tests
       run: |
         export CARDANO_MAINNET_MIRROR="$(pwd)/epochs"
+        PACKAGE="${{ matrix.package }}"
+        TEST_OPTIONS=
+        case $(cabal test "$PACKAGE" --test-options=--help) in
+          *'--[no-]color'*) TEST_OPTIONS+='--no-color';;
+          *'--color never'*) TEST_OPTIONS+='--color never';;
+        esac
```
Comment (lehins) [inline .github/workflows/haskell.yml]: I really don't like this hacky addition to our CI. Please revert this change. The fact that ImpSpec isn't playing nicely with color does not mean we should disable it for all of our tests.

Also here is a PR that fixes the problem and totally avoids the need to disable color for hspec: https://github.com/IntersectMBO/cardano-ledger/pull/4615

Also for future reference it is much easier to change config for HSpec in Haskell instead of bash :grin: 
Changing this to `ColorNever` would have disabled color output for most of hspec tests in ledger: https://github.com/IntersectMBO/cardano-ledger/blob/df783ed15f70b5b5c7a40dd5d8d62386f20bfec5/libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Common.hs#L74

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs [hunk 2]
```diff
@@ -1389,18 +1394,24 @@ impAnn msg m = do
...
+        [ annotate (color c) d
+        | (c, d) <-
+            [ (Yellow, "[")
+            , (Blue, pretty srcLocModule)
+            , (Yellow, ":")
+            , (Magenta, pretty srcLocStartLine)
+            , (Yellow, "]")
+            ]
+        ]
+    stack = vsep [prettySrcLoc' loc | (_, loc) <- getCallStack callStack]
```
Comment (lehins) [inline eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs]: This is quite a nice improvement, but I think we should also add indentation, in order to make the callstack clear and disambiguate different log entries from each other.

So, something like this would do the trick:
```suggestion
    prefix n = pretty (replicate (n - 1) ' ' <> if n == 0 then "" else "└")
    stack = vsep [prefix n <> prettySrcLoc' loc | (n, (_, loc)) <- zip [0, 2 ..] (getCallStack callStack)]
```

This will give a nice staircase:
```
       [Test.Cardano.Ledger.Shelley.ImpTest:1181]
        └[Test.Cardano.Ledger.Shelley.ImpTest:1241]
          └[Test.Cardano.Ledger.Shelley.ImpTest:1228]
            └[Test.Cardano.Ledger.Shelley.Imp.UtxoSpec:45]
```
vs
```
       [Test.Cardano.Ledger.Shelley.ImpTest:1181]
       [Test.Cardano.Ledger.Shelley.ImpTest:1241]
       [Test.Cardano.Ledger.Shelley.ImpTest:1228]
       [Test.Cardano.Ledger.Shelley.Imp.UtxoSpec:45]
```

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/UtxoSpec.hs [hunk 3]
```diff
@@ -1,6 +1,7 @@
 {-# LANGUAGE DataKinds #-}
 {-# LANGUAGE FlexibleContexts #-}
 {-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/UtxoSpec.hs:4]: What's the need for this extension? (I suspect it was an intermediate experimentation that lead to it being used)
```suggestion
```

File: eras/alonzo/impl/CHANGELOG.md [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 ## 1.10.2.1
```
Comment (lehins) [inline eras/alonzo/impl/CHANGELOG.md]: This is a visible change, so minor version needs to be bumped (in cabal file too)

Although depending on my comment about the actual instance, we might be able to keep the patch bump and remove the instance instead.
```suggestion
## 1.10.3.0
```

File: eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/TreeDiff.hs [hunk 5]
```diff
@@ -172,3 +176,17 @@ instance
         , ("pwcExUnits", toExpr pwcExUnits)
         , ("pwcCostModel", toExpr pwcCostModel)
         ]
+
+deriving instance Generic (TransactionScriptFailure era)
```
Comment (lehins) [inline eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/TreeDiff.hs]: Why is this a standalone instance instead of placed right next to the type definition?

File: eras/alonzo/impl/cardano-ledger-alonzo.cabal [hunk 6]
```diff
@@ -138,10 +138,12 @@ library testlib
         cardano-ledger-core:{cardano-ledger-core, testlib},
         cardano-ledger-shelley:{cardano-ledger-shelley, testlib},
         cardano-strict-containers,
+        plutus-core,
```
Comment (lehins) [inline eras/alonzo/impl/cardano-ledger-alonzo.cabal]: We should never depend on `plutus-core`, instead all of the types relevant to ledger can be imported through `plutus-ledger-api`
```suggestion
```

File: eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/TreeDiff.hs [hunk 7]
```diff
@@ -172,3 +176,17 @@ instance
...
+
+instance
+  ( Era era
+  , ToExpr (TxCert era)
+  , ToExpr (ContextError era)
+  , ToExpr (PlutusScript era)
+  , ToExpr (PlutusPurpose AsIx era)
+  , ToExpr (PlutusPurpose AsItem era)
+  , ToExpr EvaluationError
+  , ToExpr ExBudget
```
Comment (lehins) [inline eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/TreeDiff.hs]: These two constraints are fully saturated. Which means they cannot be satisfied.

File: eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/TreeDiff.hs [hunk 8]
```diff
@@ -172,3 +176,17 @@ instance
...
+  ) =>
+  ToExpr (TransactionScriptFailure era)
```
Comment (lehins) [inline eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/TreeDiff.hs]: So, why did you need this instance?

Currently it is broken, so you could not have used it in any of the tests, therefore I suspect it can be removed.

On the other hand if we really do need it then we need to define orphan `ToExpr` instances for `EvaluationError` and `ExBudget`. However, those would have to be placed into `cardano-ledgre-core:testlib`

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs [hunk 9]
```diff
@@ -1389,18 +1394,24 @@ impAnn msg m = do
...
+
+-- | Adds a Doc to the log, which is only shown if the test fails
+logEntry :: HasCallStack => Doc AnsiStyle -> ImpTestM era ()
+logEntry = logWithCallStack ?callStack
```
Comment (lehins) [inline eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs]: Let's keep the old version that uses `String`. It is not always convenient to work with `Doc`
```suggestion
logDocEntry :: HasCallStack => Doc AnsiStyle -> ImpTestM era ()
logDocEntry = logWithCallStack ?callStack

logEntry :: HasCallStack => String -> ImpTestM era ()
logEntry = logWithCallStack ?callStack . fromString
```

File: eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/Serialisation/Golden/Encoding.hs [hunk 10]
```diff
@@ -1135,11 +1136,12 @@ tests =
               ]
        in testCase "ledger state golden test" $
             unless (actual == expected) $
-              assertFailure $
-                unlines
-                  [ "Expected: " ++ show expectedHex
-                  , "Actual: " ++ show actualHex
-                  , diffExpr (CBORBytes expected) (CBORBytes actual)
+              assertFailure . ansiDocToString $
```
Comment (lehins) [inline eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/Serialisation/Golden/Encoding.hs:1139]: tasty doesn't play well with color, but we can keep it this way, since we'll need to move this test case to the newer test suite that uses hspec anyways.

File: libs/cardano-ledger-binary/CHANGELOG.md [hunk 11]
```diff
@@ -1,8 +1,18 @@
...
-## 1.3.4.1
+## 1.4.0.0
 
-*
+### `testlib`
+
+* Re-export types `Doc` and `AnsiStyle` in `Test.Cardano.Ledger.Binary.TreeDiff`
+* Rename functions:
+  - `diffExprNoColor` -> `diffExprString`
+  - `diffExprNoColor` -> `diffExprCompactString`
```
Comment (lehins) [inline libs/cardano-ledger-binary/CHANGELOG.md]: I don't think you can rename one function into two different ones :smile: 

It should be.
* Remove ...
* Add ...

Renaming is only applicable when neither the type signature nor the functionality have experienced any change

File: libs/cardano-ledger-binary/CHANGELOG.md [hunk 12]
```diff
@@ -1,8 +1,18 @@
 # Version history for `cardano-ledger-binary`
 
-## 1.3.4.1
+## 1.4.0.0
 
-*
+### `testlib`
+
```
Comment (lehins) [inline libs/cardano-ledger-binary/CHANGELOG.md:6]:
```suggestion
* `diffExpr` changed type signature
* Add `diffExprString`, which replaces old implementation of `diffExpr`
```

File: libs/cardano-ledger-binary/testlib/Test/Cardano/Ledger/Binary/TreeDiff.hs [hunk 13]
```diff
@@ -54,6 +65,14 @@ import Test.Tasty.HUnit (Assertion, assertFailure)
 trimExprViaShow :: Show a => Int -> a -> Expr
 trimExprViaShow _n x = defaultExprViaShow x -- App (take n (drop 1 (show x)) ++ "..") []
 
+tableDoc :: Maybe (Doc AnsiStyle) -> [(String, Doc AnsiStyle)] -> Doc AnsiStyle
+tableDoc mTitle rows =
```
Comment (lehins) [inline libs/cardano-ledger-binary/testlib/Test/Cardano/Ledger/Binary/TreeDiff.hs:69]: `maximum` is a partial function that won't give you much hint of where the exception came from.
```suggestion
tableDoc :: HasCallStack => Maybe (Doc AnsiStyle) -> [(String, Doc AnsiStyle)] -> Doc AnsiStyle
tableDoc mTitle [] = error "`tableDoc` expects a non-empty list"
tableDoc mTitle rows =
```
Alternatively you could require `NonEmpty` instead of a list. More type safety :wink: 
Feel free to decide on which approach to choose with this helper test function

File: eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/Imp/UtxowSpec.hs [hunk 14]
```diff
@@ -63,7 +64,8 @@ spec = describe "UTXOW" $ do
   it "ExtraRedeemers/RedeemerPointerPointsToNothing" $
     -- There is ExtraRedeemers test for PlutusV1 in Alonzo, thus we start with PlutusV2
     forM_ ([PlutusV2 .. eraMaxLanguage @era] :: [Language]) $ \lang -> do
-      logEntry $ "Testing for " ++ show lang
+      logEntry $ "Testing for " <> viaShow lang
+      logEntry $ "Testing for " <> viaShow lang
```
Comment (lehins) [inline eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/Imp/UtxowSpec.hs]: A bit of duplication. I've also suggested keeping type signature of `logEntry` unchanged, so we wouldn't have to change this kind of occurrences
```suggestion
      logEntry $ "Testing for " <> show lang
```

File: eras/shelley/impl/CHANGELOG.md [hunk 15]
```diff
@@ -11,6 +11,7 @@
 ### `testlib`
 
 * Remove `minMajorPV` and `maxMajorPV` from `Constants`
+* Change `logEntry` to take a `Doc AnsiStyle` instead of a `String`
```
Comment (lehins) [inline eras/shelley/impl/CHANGELOG.md]:
```suggestion
* Add `logDocEntry` that takes a `Doc AnsiStyle` instead of a `String`
```

File: libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Common.hs [hunk 16]
```diff
@@ -8,8 +8,12 @@ module Test.Cardano.Ledger.Common (
   -- * Expr
   ToExpr (..),
   showExpr,
+  ansiExpr,
+  ansiExprString,
   diffExpr,
+  diffExprString,
```
Comment (lehins) [inline libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Common.hs:14]: Feel free to keep the names the way you've done them, if you'd like. But, I just wanted to make a comment for the future.

I would have added instead a new function, named something like `diffDocExpr`, while keeping old one `diffExpr` intact. That would have resulted in a lot less API breakage. (Same comment applies to changes of functions in this PR)

It is always good practice to keep breakage to a minimum. In this case it is just test helper functions, which probably don't have much impact downstream. But for regular ledger API this comment is very much applicable.

**Review by lehins — CHANGES_REQUESTED** [review]
This is awesome! Thank you for making all those improvements.

I've implemented the missing piece in #4615. I've based it on this PR and targeted this branch for now.

Let's get this one cleaned up and merged and I'll re-target mine onto master after this one is merged.

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/UtxoSpec.hs (hunk 3 above)
Comment (neilmayhew) [reply eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/UtxoSpec.hs:4]: The module doesn't compile without it. It's needed because `logEntry` was changed to take a `Doc`.

I expect you see this as a further argument for introducing a new name rather than changing the type :smile:

File: eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/TreeDiff.hs (hunk 5 above)
Comment (neilmayhew) [reply eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/TreeDiff.hs]: It's needed only for tests so I didn't think it was necessary to add it to the non-test module where the type is defined.

Comment (neilmayhew) [reply eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/TreeDiff.hs]: However, I've taken it out now, because it's not needed as you pointed out

File: libs/cardano-ledger-binary/testlib/Test/Cardano/Ledger/Binary/TreeDiff.hs (hunk 13 above)
Comment (neilmayhew) [reply libs/cardano-ledger-binary/testlib/Test/Cardano/Ledger/Binary/TreeDiff.hs:69]: I've used `foldr max 0` instead, and `hlint` showed me I could fuse that with the `map (length . fst)` to give `foldr (max . length . fst) 0`

File: eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/TreeDiff.hs (hunk 8 above)
Comment (neilmayhew) [reply eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/TreeDiff.hs]: I originally added it so I could use `ansiExpr` instead of `viaShow` [here](https://github.com/IntersectMBO/cardano-ledger/blob/e34138b741bb5bace2949b19b74ed2aa3cd83ff1/eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/ImpTest.hs#L213). However, the same module then needed `ToExpr (PlutusPurpose AsIx era)` so I gave up and just used `viaShow` (but didn't take out the instance right away in case I wanted to go back to it).

File: eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/ImpTest.hs [hunk 17]
```diff
@@ -209,7 +210,7 @@ fixupRedeemers tx = impAnn "fixupRedeemers" $ do
   exUnitsPerPurpose <-
     fmap (Map.mapMaybe id) $ forM reports $ \case
       Left err -> do
-        logEntry $ "Execution Units estimation error: " ++ show err
+        logDocEntry $ "Execution Units estimation error: " <> viaShow err
```
Comment (lehins) [inline eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/ImpTest.hs]: Let's reduce the diff
```suggestion
        logEntry $ "Execution Units estimation error: " <> show err
```

File: eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/ImpTest.hs [hunk 18]
```diff
@@ -221,7 +222,7 @@ fixupRedeemers tx = impAnn "fixupRedeemers" $ do
             Nothing ->
               case Map.lookup ptr exUnitsPerPurpose of
                 Nothing -> do
-                  logEntry $ "Missing Redeemer Ptr from execution estimation: " ++ show ptr
+                  logDocEntry $ "Missing Redeemer Ptr from execution estimation: " <> viaShow ptr
```
Comment (lehins) [inline eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/ImpTest.hs]:
```suggestion
                  logEntry $ "Missing Redeemer Ptr from execution estimation: " <> show ptr
```

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs [hunk 19]
```diff
@@ -1293,11 +1310,11 @@ electBasicCommittee ::
     , GovPurposeId 'CommitteePurpose era
     )
 electBasicCommittee = do
-  logEntry "Setting up a DRep"
+  logDocEntry "Setting up a DRep"
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs]:
```suggestion
  logEntry "Setting up a DRep"
```

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs [hunk 20]
```diff
@@ -1293,11 +1310,11 @@ electBasicCommittee ::
...
   (drep, _, _) <- setupSingleDRep 1_000_000
   (spoC, _, _) <- setupPoolWithStake $ Coin 1_000_000
 
-  logEntry "Registering committee member"
+  logDocEntry "Registering committee member"
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs]:
```suggestion
  logEntry "Registering committee member"
```

File: libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp/Ratify.hs [hunk 21]
```diff
@@ -77,7 +77,7 @@ spec = describe "RATIFY" . withImpStateWithProtVer @Conway (natVersion @10) $ do
         ratSt
         (RatifySignal (noConfidenceGAS SSeq.:<| SSeq.Empty))
   it "Duplicate CC hot keys count as separate votes" $ do
-    logEntry "Setting up a DRep"
+    logDocEntry "Setting up a DRep"
```
Comment (lehins) [inline libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp/Ratify.hs]:
```suggestion
    logEntry "Setting up a DRep"
```

File: libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp/Ratify.hs [hunk 22]
```diff
@@ -90,7 +90,7 @@ spec = describe "RATIFY" . withImpStateWithProtVer @Conway (natVersion @10) $ do
     SJust (Committee {committeeMembers = oldCommittee}) <-
       getsNES $ nesEsL . epochStateGovStateL . committeeGovStateL
 
-    logEntry "Electing the committee"
+    logDocEntry "Electing the committee"
```
Comment (lehins) [inline libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp/Ratify.hs]:
```suggestion
    logEntry "Electing the committee"
```

File: libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp/Ratify.hs [hunk 23]
```diff
@@ -118,17 +118,17 @@ spec = describe "RATIFY" . withImpStateWithProtVer @Conway (natVersion @10) $ do
       passNEpochs 2
       getLastEnactedCommittee `shouldReturn` SJust (GovPurposeId committeeId)
 
-    logEntry "Registering hotkeys"
+    logDocEntry "Registering hotkeys"
```
Comment (lehins) [inline libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp/Ratify.hs]:
```suggestion
    logEntry "Registering hotkeys"
```

File: libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp/Ratify.hs [hunk 24]
```diff
@@ -118,17 +118,17 @@ spec = describe "RATIFY" . withImpStateWithProtVer @Conway (natVersion @10) $ do
...
     _ <- registerCommitteeHotKeys (pure hotKey) (ccCold0 NE.:| [ccCold1])
 
-    logEntry "Proposing a new constitution"
+    logDocEntry "Proposing a new constitution"
```
Comment (lehins) [inline libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp/Ratify.hs]:
```suggestion
    logEntry "Proposing a new constitution"
```

File: libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp/Ratify.hs [hunk 25]
```diff
@@ -118,17 +118,17 @@ spec = describe "RATIFY" . withImpStateWithProtVer @Conway (natVersion @10) $ do
...
     newConstitution <- arbitrary
     constitutionId <- submitGovAction $ NewConstitution SNothing newConstitution
     submitYesVote_ (CommitteeVoter hotKey) constitutionId
     submitYesVote_ (DRepVoter credDRep) constitutionId
     constitutionGAS <- getGovActionState constitutionId
 
-    logEntry "Testing conformance"
+    logDocEntry "Testing conformance"
```
Comment (lehins) [inline libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp/Ratify.hs]:
```suggestion
    logEntry "Testing conformance"
```

File: libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp/Ratify.hs [hunk 26]
```diff
@@ -142,20 +142,20 @@ spec = describe "RATIFY" . withImpStateWithProtVer @Conway (natVersion @10) $ do
         ratEnv
         ratSt
         ratSig
-    logEntry "===context==="
+    logDocEntry "===context==="
```
Comment (lehins) [inline libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp/Ratify.hs]:
```suggestion
    logEntry "===context==="
```

File: libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp/Ratify.hs [hunk 27]
```diff
@@ -142,20 +142,20 @@ spec = describe "RATIFY" . withImpStateWithProtVer @Conway (natVersion @10) $ do
...
     logToExpr execCtx
-    logEntry "===environment==="
+    logDocEntry "===environment==="
```
Comment (lehins) [inline libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp/Ratify.hs]:
```suggestion
    logEntry "===environment==="
```

File: libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp/Ratify.hs [hunk 28]
```diff
@@ -142,20 +142,20 @@ spec = describe "RATIFY" . withImpStateWithProtVer @Conway (natVersion @10) $ do
...
     logToExpr ratEnv
-    logEntry "===state==="
+    logDocEntry "===state==="
```
Comment (lehins) [inline libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp/Ratify.hs]:
```suggestion
    logEntry "===state==="
```

File: libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp/Ratify.hs [hunk 29]
```diff
@@ -142,20 +142,20 @@ spec = describe "RATIFY" . withImpStateWithProtVer @Conway (natVersion @10) $ do
...
     logToExpr ratSt
-    logEntry "===signal==="
+    logDocEntry "===signal==="
```
Comment (lehins) [inline libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp/Ratify.hs]:
```suggestion
    logEntry "===signal==="
```

File: libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp/Ratify.hs [hunk 30]
```diff
@@ -142,20 +142,20 @@ spec = describe "RATIFY" . withImpStateWithProtVer @Conway (natVersion @10) $ do
...
     logToExpr ratSig
-    logEntry "Impl res:"
+    logDocEntry "Impl res:"
```
Comment (lehins) [inline libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp/Ratify.hs]:
```suggestion
    logEntry "Impl res:"
```

File: libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp/Ratify.hs [hunk 31]
```diff
@@ -142,20 +142,20 @@ spec = describe "RATIFY" . withImpStateWithProtVer @Conway (natVersion @10) $ do
...
     logToExpr implRes
-    logEntry "Agda res:"
+    logDocEntry "Agda res:"
```
Comment (lehins) [inline libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp/Ratify.hs]:
```suggestion
    logEntry "Agda res:"
```

File: libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp/Ratify.hs [hunk 32]
```diff
@@ -142,20 +142,20 @@ spec = describe "RATIFY" . withImpStateWithProtVer @Conway (natVersion @10) $ do
...
     logToExpr agdaRes
-    logEntry "Extra information:"
-    logEntry $
+    logDocEntry "Extra information:"
```
Comment (lehins) [inline libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp/Ratify.hs]:
```suggestion
    logEntry "Extra information:"
```

**Review by lehins — APPROVED** [review]
Looks great!
Definitely a really nice quality of life improvement.

File: eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/ImpTest.hs (hunk 17 above)
Comment (neilmayhew) [reply eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/ImpTest.hs]: Sure, I can do that. The reason I didn't do it before is that I think it reads better when there's only, or mainly, one log function used in the file, and only one imported at the top. The cognitive load of switching back and forth between one function and the other depending on exactly what's being printed seems unnecessary when `OverloadedStrings` can take care of it for you.

Comment (lehins) [reply eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/ImpTest.hs]: > The reason I didn't do it before is that I think it reads better when there's only, or mainly, one log function used in the file

That goes against having two different functions. You can't force users to use only one specific function in a module.

>  and only one imported at the top

Both are in scope by default. We don't do explicit imports on imp functionality

> The cognitive load of switching back and forth between one function and the other depending on exactly what's being printed seems unnecessary when OverloadedStrings can take care of it for you.

You have to know to use `viaShow` vs `show` even in presence of `OverloadedString`. Most importantly, we now have two slightly different functions. It is absolutely ok to use either one of them at any point. Those are not the type of functions that bring cognitive overhead, they are dead simple logging functions.

Please, revert them back. I really don't like writing long explanation for what used to be a one line suggestion.

File: eras/alonzo/impl/cardano-ledger-alonzo.cabal [hunk 33]
```diff
@@ -1,6 +1,6 @@
 cabal-version:      3.0
 name:               cardano-ledger-alonzo
-version:            1.10.2.1
+version:            1.10.2.0
```
Comment (lehins) [inline eras/alonzo/impl/cardano-ledger-alonzo.cabal]: Why are you downgrading the version to the one that was released?
```suggestion
version:            1.10.2.1
```

File: eras/alonzo/impl/CHANGELOG.md [hunk 34]
```diff
@@ -1,9 +1,5 @@
 # Version history for `cardano-ledger-alonzo`
 
-## 1.10.2.1
```
Comment (lehins) [inline eras/alonzo/impl/CHANGELOG.md]: New sections are never added, but they are also never removed

**Review by lehins — CHANGES_REQUESTED** [review]

File: eras/alonzo/impl/cardano-ledger-alonzo.cabal (hunk 33 above)
Comment (neilmayhew) [reply eras/alonzo/impl/cardano-ledger-alonzo.cabal]: Thanks for catching this. I'm not sure why I'm having such a hard time getting this right.

File: eras/shelley/impl/CHANGELOG.md [hunk 35]
```diff
@@ -12,6 +12,7 @@
 ### `testlib`
 
 * Remove `minMajorPV` and `maxMajorPV` from `Constants`
+* Add `logDocEntry` that takes a `Doc AnsiStyle` instead of a `String`
```
Comment (lehins) [inline eras/shelley/impl/CHANGELOG.md]:
```suggestion
* Add `logDoc` that takes a `Doc AnsiStyle` instead of a `String`
* Rename `logEntry` to `logString`
```

**Review by lehins — APPROVED** [review]
Thank you for doing the `logDoc` and `logString` renames. That of course also needs a changelog :slightly_smiling_face:
